File: task_5/main.py
import logging
import sys
import random

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[
        logging.StreamHandler(sys.stdout),
        logging.FileHandler('triangle_path.log', mode='w', encoding='utf-8')
    ]
)

class LogMessages:
    # Algorithm operation messages
    ALGORITHM_START = "Starting minimum path sum calculation for triangle"
    ALGORITHM_COMPLETE = "Algorithm complete. Minimum path sum: {}"
    ALGORITHM_EMPTY_INPUT = "Empty triangle provided"

    # Triangle processing messages
    TRIANGLE_SIZE = "Processing triangle with {} rows"
    ROW_PROCESSING = "Processing row {} with {} elements"
    DP_INITIALIZATION = "Initializing DP with base row: {}"
    DP_UPDATE = "Updating dp[{}] = min({} + {}, {} + {}) = {}"

    # Path reconstruction messages
    PATH_RECONSTRUCTION_START = "Starting path reconstruction"
    PATH_ELEMENT_ADDED = "Added element {} at position {} to path"
    PATH_COMPLETE = "Minimum path: {}"

    # Test messages
    TEST_START = "Starting test case {}"
    TEST_RESULT = "Test {}: Expected sum = {}, Got = {} | Expected path = {}, Got = {}"
    TEST_PASSED = "Test {} PASSED"
    TEST_FAILED = "Test {} FAILED"

    # Generation messages
    GENERATION_START = "Generating random test case {} with {} rows"
    GENERATION_COMPLETE = "Generated triangle: {}"

def minimum_total(triangle):
    """
    Finds the minimum path sum from top to bottom of the triangle.

    Args:
        triangle: List of lists representing the triangle

    Returns:
        tuple: (minimum_sum, path_taken)
    """
    try:
        logging.info(LogMessages.ALGORITHM_START)

        if not triangle or not triangle[0]:
            logging.warning(LogMessages.ALGORITHM_EMPTY_INPUT)
            return 0, []

        n = len(triangle)
        logging.info(LogMessages.TRIANGLE_SIZE.format(n))

        # Create DP table
        dp = [[0] * len(triangle[i]) for i in range(n)]

        # Initialize last row
        for j in range(len(triangle[n-1])):
            dp[n-1][j] = triangle[n-1][j]

        logging.info(LogMessages.DP_INITIALIZATION.format(dp[n-1]))

        # Fill DP table bottom-up
        for i in range(n-2, -1, -1):
            # logging.info(LogMessages.ROW_PROCESSING.format(i, len(triangle[i])))

            for j in range(len(triangle[i])):
                dp[i][j] = triangle[i][j] + min(dp[i+1][j], dp[i+1][j+1])
                logging.debug(LogMessages.DP_UPDATE.format(
                    j, triangle[i][j], dp[i+1][j], triangle[i][j], dp[i+1][j+1], dp[i][j]
                ))

        # Reconstruct path
        path = reconstruct_path_correct(triangle, dp)
        min_sum = dp[0][0]

        logging.info(LogMessages.ALGORITHM_COMPLETE.format(min_sum))
        logging.info(LogMessages.PATH_COMPLETE.format(" -> ".join(map(str, path))))

        return min_sum, path

    except Exception as error:
        logging.error(f"Error in minimum path calculation: {str(error)}")
        return float('inf'), []

def reconstruct_path_correct(triangle, dp):
    """
    Correctly reconstructs the path using the DP table.

    Args:
        triangle: The original triangle
        dp: The DP table filled with minimum sums

    Returns:
        list: The path from top to bottom
    """
    try:
        logging.info(LogMessages.PATH_RECONSTRUCTION_START)

        n = len(triangle)
        path = []

        current_col = 0
        path.append(triangle[0][current_col])
        # logging.info(LogMessages.PATH_ELEMENT_ADDED.format(triangle[0][current_col], current_col))

        # Reconstruct top-down path
        for i in range(1, n):
            # Check which path gives the current DP value
            # The correct next step is the one that matches: dp[i][col] == dp[i-1][current_col] - triangle[i-1][current_col]
            expected_value = dp[i-1][current_col] - triangle[i-1][current_col]

            if dp[i][current_col] == expected_value:
                path.append(triangle[i][current_col])
                # logging.info(LogMessages.PATH_ELEMENT_ADDED.format(triangle[i][current_col], current_col))
            else:
                # Move to next column
                current_col += 1
                path.append(triangle[i][current_col])
                # logging.info(LogMessages.PATH_ELEMENT_ADDED.format(triangle[i][current_col], current_col))

        return path

    except Exception as error:
        logging.error(f"Error in path reconstruction: {str(error)}")
        return []

class TriangleTests:
    """
    Test cases for the triangle path algorithm.
    """

    @staticmethod
    def get_basic_tests():
        """Returns basic test cases from the problem statement."""
        return [
            {
                "name": "Basic Triangle 1",
                "triangle": [[2], [3, 4], [6, 5, 7], [4, 1, 8, 3]],
                "expected_sum": 11,
                "expected_path": [2, 3, 5, 1]
            },
            {
                "name": "Basic Triangle 2",
                "triangle": [[-1], [2, 3], [1, -1, -3], [4, 2, 1, 3]],
                "expected_sum": 0,
                "expected_path": [-1, 3, -3, 1]
            }
        ]

    @staticmethod
    def get_edge_tests():
        """Returns edge case tests."""
        return [
            {
                "name": "Single Element",
                "triangle": [[5]],
                "expected_sum": 5,
                "expected_path": [5]
            },
            {
                "name": "Two Rows",
                "triangle": [[1], [2, 3]],
                "expected_sum": 3,
                "expected_path": [1, 2]
            },
            {
                "name": "All Same Values",
                "triangle": [[1], [1, 1], [1, 1, 1]],
                "expected_sum": 3,
                "expected_path": [1, 1, 1]
            },
            {
                "name": "Negative Values",
                "triangle": [[-1], [-2, -3], [-4, -5, -6]],
                "expected_sum": -10,
                "expected_path": [-1, -3, -6]
            }
        ]

    @staticmethod
    def get_large_tests():
        """Returns larger test cases."""
        return [
            {
                "name": "5x5 Triangle",
                "triangle": [
                    [1],
                    [2, 3],
                    [4, 5, 6],
                    [7, 8, 9, 10],
                    [11, 12, 13, 14, 15]
                ],
                "expected_sum": 1 + 2 + 4 + 7 + 11,
                "expected_path": [1, 2, 4, 7, 11]
            }
        ]

class TriangleGenerator:
    """
    Generates random triangles for testing.
    """

    @staticmethod
    def generate_random_triangle(rows, min_val=-10, max_val=10):
        """
        Generates a random triangle with given number of rows.

        Args:
            rows: Number of rows in the triangle
            min_val: Minimum value for elements
            max_val: Maximum value for elements

        Returns:
            list: Generated triangle
        """
        logging.info(LogMessages.GENERATION_START.format(rows, rows))

        triangle = []
        for i in range(rows):
            row = [random.randint(min_val, max_val) for _ in range(i + 1)]
            triangle.append(row)

        logging.info(LogMessages.GENERATION_COMPLETE.format(triangle))
        return triangle

    @staticmethod
    def generate_positive_triangle(rows, max_val=20):
        """Generates triangle with only positive values."""
        return TriangleGenerator.generate_random_triangle(rows, 1, max_val)

    @staticmethod
    def generate_negative_triangle(rows, min_val=-20):
        """Generates triangle with only negative values."""
        return TriangleGenerator.generate_random_triangle(rows, min_val, -1)

    @staticmethod
    def generate_mixed_triangle(rows):
        """Generates triangle with mixed positive and negative values."""
        return TriangleGenerator.generate_random_triangle(rows, -15, 15)

def run_test(test_case, test_number):
    """
    Runs a single test case and returns results.

    Args:
        test_case: Dictionary with test case data
        test_number: Test case number

    Returns:
        tuple: (passed, actual_sum, actual_path)
    """
    logging.info(LogMessages.TEST_START.format(test_number))

    triangle = test_case["triangle"]
    expected_sum = test_case["expected_sum"]
    expected_path = test_case["expected_path"]

    actual_sum, actual_path = minimum_total(triangle)

    sum_correct = actual_sum == expected_sum
    path_correct = actual_path == expected_path

    passed = sum_correct and path_correct

    if passed:
        logging.info(LogMessages.TEST_PASSED.format(test_number))
    else:
        logging.info(LogMessages.TEST_FAILED.format(test_number))

    logging.info(LogMessages.TEST_RESULT.format(
        test_number, expected_sum, actual_sum,
        " -> ".join(map(str, expected_path)),
        " -> ".join(map(str, actual_path))
    ))

    return passed, actual_sum, actual_path

def run_test_suite():
    """
    Runs all test suites and returns summary.
    """
    logging.info("RUNNING COMPREHENSIVE TEST SUITE")

    all_tests = []
    results = []

    # Basic tests
    basic_tests = TriangleTests.get_basic_tests()
    all_tests.extend(basic_tests)

    # Edge tests
    edge_tests = TriangleTests.get_edge_tests()
    all_tests.extend(edge_tests)

    # Large tests
    large_tests = TriangleTests.get_large_tests()
    all_tests.extend(large_tests)

    # Run all predefined tests
    for i, test_case in enumerate(all_tests):
        passed, actual_sum, actual_path = run_test(test_case, i + 1)
        results.append((test_case["name"], passed, actual_sum, actual_path))

    # Generate random 3 tests
    random_tests = []
    for i in range(3):
        rows = random.randint(3, 6)
        triangle = TriangleGenerator.generate_random_triangle(rows)

        expected_sum, expected_path = minimum_total(triangle)

        random_test = {
            "name": f"Random Test {i + 1}",
            "triangle": triangle,
            "expected_sum": expected_sum,
            "expected_path": expected_path
        }
        random_tests.append(random_test)

    # Run random tests (should always pass)
    start_idx = len(all_tests)
    for i, test_case in enumerate(random_tests):
        passed, actual_sum, actual_path = run_test(test_case, start_idx + i + 1)
        results.append((test_case["name"], passed, actual_sum, actual_path))

    return results

def print_test_summary(results):
    """
    Prints a summary of test results.

    Args:
        results: List of test results
    """
    logging.info("TEST SUMMARY")

    passed_count = sum(1 for _, passed, _, _ in results if passed)
    total_count = len(results)

    logging.info(f"Total Tests: {total_count}")
    logging.info(f"Passed: {passed_count}")
    logging.info(f"Failed: {total_count - passed_count}")
    logging.info(f"Success Rate: {passed_count/total_count*100:.1f}%")

    logging.info("Detailed Results:")
    for name, passed, actual_sum, actual_path in results:
        status = "PASS" if passed else "FAIL"
        logging.info(f"{name:<20} {status:<6} Sum: {actual_sum:<6} Path: {' -> '.join(map(str, actual_path))}")

def benchmark_algorithm():
    """
    Benchmarks the algorithm with large triangles.
    """
    logging.info("BENCHMARK WITH LARGE TRIANGLES")

    import time

    sizes = [10, 20, 50, 100]

    for size in sizes:
        triangle = TriangleGenerator.generate_random_triangle(size, -100, 100)

        start_time = time.time()
        min_sum, path = minimum_total(triangle)
        end_time = time.time()

        execution_time = end_time - start_time
        logging.info(f"Size: {size}, Time: {execution_time:.4f}s, Min Sum: {min_sum}")
        logging.info(f"Size {size}: {execution_time:.4f}s (sum: {min_sum})")

def main():
    """
    Main execution function.
    """
    try:
        # Run comprehensive test suite
        test_results = run_test_suite()

        # Print summary
        print_test_summary(test_results)

        # Run BENCHMARK
        logging.info(f"BENCHMARK ... ")
        benchmark_algorithm()

        original_tests = TriangleTests.get_basic_tests()
        for test in original_tests:
            min_sum, path = minimum_total(test["triangle"])

    except Exception as error:
        logging.error(f"Error in main execution: {str(error)}")

if __name__ == "__main__":
    main()

File: task_5/main.cpp
#include <iostream>
#include <vector>
#include <string>
#include <sstream>
#include <algorithm>
#include <limits>
#include <random>
#include <chrono>
#include <fstream>
#include <memory>

class Logger {
private:
    std::ofstream file_stream;
    bool log_to_file;
    
public:
    Logger(bool to_file = true) : log_to_file(to_file) {
        if (log_to_file) {
            file_stream.open("triangle_path.log", std::ios::out);
        }
    }
    
    ~Logger() {
        if (file_stream.is_open()) {
            file_stream.close();
        }
    }
    
    void info(const std::string& message) {
        log("INFO", message);
    }
    
    void warning(const std::string& message) {
        log("WARNING", message);
    }
    
    void error(const std::string& message) {
        log("ERROR", message);
    }
    
    void debug(const std::string& message) {
        log("DEBUG", message);
    }
    
private:
    void log(const std::string& level, const std::string& message) {
        auto now = std::chrono::system_clock::now();
        auto time_t = std::chrono::system_clock::to_time_t(now);
        
        std::stringstream ss;
        ss << std::ctime(&time_t);
        std::string time_str = ss.str();
        time_str.pop_back();
        
        std::string log_message = time_str + " - " + level + " - " + message;
        
        std::cout << log_message << std::endl;
        if (log_to_file && file_stream.is_open()) {
            file_stream << log_message << std::endl;
        }
    }
};

struct LogMessages {
    static const std::string ALGORITHM_START;
    static const std::string ALGORITHM_COMPLETE;
    static const std::string ALGORITHM_EMPTY_INPUT;
    static const std::string TRIANGLE_SIZE;
    static const std::string ROW_PROCESSING;
    static const std::string DP_INITIALIZATION;
    static const std::string DP_UPDATE;
    static const std::string PATH_RECONSTRUCTION_START;
    static const std::string PATH_ELEMENT_ADDED;
    static const std::string PATH_COMPLETE;
    static const std::string TEST_START;
    static const std::string TEST_RESULT;
    static const std::string TEST_PASSED;
    static const std::string TEST_FAILED;
    static const std::string GENERATION_START;
    static const std::string GENERATION_COMPLETE;
};

const std::string LogMessages::ALGORITHM_START = "Starting minimum path sum calculation for triangle";
const std::string LogMessages::ALGORITHM_COMPLETE = "Algorithm complete. Minimum path sum: ";
const std::string LogMessages::ALGORITHM_EMPTY_INPUT = "Empty triangle provided";
const std::string LogMessages::TRIANGLE_SIZE = "Processing triangle with ";
const std::string LogMessages::ROW_PROCESSING = "Processing row ";
const std::string LogMessages::DP_INITIALIZATION = "Initializing DP with base row: ";
const std::string LogMessages::DP_UPDATE = "Updating dp[";
const std::string LogMessages::PATH_RECONSTRUCTION_START = "Starting path reconstruction";
const std::string LogMessages::PATH_ELEMENT_ADDED = "Added element ";
const std::string LogMessages::PATH_COMPLETE = "Minimum path: ";
const std::string LogMessages::TEST_START = "Starting test case ";
const std::string LogMessages::TEST_RESULT = "Test ";
const std::string LogMessages::TEST_PASSED = "Test PASSED";
const std::string LogMessages::TEST_FAILED = "Test FAILED";
const std::string LogMessages::GENERATION_START = "Generating random test case ";
const std::string LogMessages::GENERATION_COMPLETE = "Generated triangle: ";

std::string vectorToString(const std::vector<int>& vec) {
    std::stringstream ss;
    ss << "[";
    for (size_t i = 0; i < vec.size(); ++i) {
        ss << vec[i];
        if (i < vec.size() - 1) ss << ", ";
    }
    ss << "]";
    return ss.str();
}

std::string triangleToString(const std::vector<std::vector<int>>& triangle) {
    std::stringstream ss;
    ss << "[";
    for (size_t i = 0; i < triangle.size(); ++i) {
        ss << "[";
        for (size_t j = 0; j < triangle[i].size(); ++j) {
            ss << triangle[i][j];
            if (j < triangle[i].size() - 1) ss << ", ";
        }
        ss << "]";
        if (i < triangle.size() - 1) ss << ", ";
    }
    ss << "]";
    return ss.str();
}

std::string pathToString(const std::vector<int>& path) {
    std::stringstream ss;
    for (size_t i = 0; i < path.size(); ++i) {
        ss << path[i];
        if (i < path.size() - 1) ss << " -> ";
    }
    return ss.str();
}

std::pair<int, std::vector<int>> minimum_total(const std::vector<std::vector<int>>& triangle, Logger& logger) {
    try {
        logger.info(LogMessages::ALGORITHM_START);

        if (triangle.empty() || triangle[0].empty()) {
            logger.warning(LogMessages::ALGORITHM_EMPTY_INPUT);
            return {0, {}};
        }

        int n = triangle.size();
        logger.info(LogMessages::TRIANGLE_SIZE + std::to_string(n) + " rows");

        
        std::vector<std::vector<int>> dp(n);
        for (int i = 0; i < n; ++i) {
            dp[i].resize(triangle[i].size());
        }

        for (size_t j = 0; j < triangle[n-1].size(); ++j) {
            dp[n-1][j] = triangle[n-1][j];
        }

        logger.info(LogMessages::DP_INITIALIZATION + vectorToString(dp[n-1]));

        for (int i = n-2; i >= 0; --i) {
            for (size_t j = 0; j < triangle[i].size(); ++j) {
                dp[i][j] = triangle[i][j] + std::min(dp[i+1][j], dp[i+1][j+1]);
                
                std::string debug_msg = LogMessages::DP_UPDATE + 
                    std::to_string(j) + "] = min(" + 
                    std::to_string(triangle[i][j]) + " + " + std::to_string(dp[i+1][j]) + ", " +
                    std::to_string(triangle[i][j]) + " + " + std::to_string(dp[i+1][j+1]) + ") = " +
                    std::to_string(dp[i][j]);
                logger.debug(debug_msg);
            }
        }

        logger.info(LogMessages::PATH_RECONSTRUCTION_START);
        std::vector<int> path;
        int current_col = 0;
        path.push_back(triangle[0][current_col]);

        for (int i = 1; i < n; ++i) {
            int expected_value = dp[i-1][current_col] - triangle[i-1][current_col];
            
            if (dp[i][current_col] == expected_value) {
                path.push_back(triangle[i][current_col]);
            } else {
                current_col += 1;
                path.push_back(triangle[i][current_col]);
            }
        }

        int min_sum = dp[0][0];
        logger.info(LogMessages::ALGORITHM_COMPLETE + std::to_string(min_sum));
        logger.info(LogMessages::PATH_COMPLETE + pathToString(path));

        return {min_sum, path};

    } catch (const std::exception& error) {
        logger.error("Error in minimum path calculation: " + std::string(error.what()));
        return {std::numeric_limits<int>::max(), {}};
    }
}

class TriangleGenerator {
private:
    std::random_device rd;
    std::mt19937 gen;
    
public:
    TriangleGenerator() : gen(rd()) {}
    
    std::vector<std::vector<int>> generate_random_triangle(int rows, int min_val = -10, int max_val = 10) {
        Logger logger(false);
        logger.info(LogMessages::GENERATION_START + std::to_string(rows) + " with " + std::to_string(rows) + " rows");
        
        std::vector<std::vector<int>> triangle;
        std::uniform_int_distribution<int> dis(min_val, max_val);
        
        for (int i = 0; i < rows; ++i) {
            std::vector<int> row;
            for (int j = 0; j <= i; ++j) {
                row.push_back(dis(gen));
            }
            triangle.push_back(row);
        }
        
        logger.info(LogMessages::GENERATION_COMPLETE + triangleToString(triangle));
        return triangle;
    }
};

struct TestCase {
    std::string name;
    std::vector<std::vector<int>> triangle;
    int expected_sum;
    std::vector<int> expected_path;
};

class TriangleTests {
public:
    static std::vector<TestCase> get_basic_tests() {
        return {
            {"Basic Triangle 1", {{2}, {3, 4}, {6, 5, 7}, {4, 1, 8, 3}}, 11, {2, 3, 5, 1}},
            {"Basic Triangle 2", {{-1}, {2, 3}, {1, -1, -3}, {4, 2, 1, 3}}, 0, {-1, 3, -3, 1}}
        };
    }
    
    static std::vector<TestCase> get_edge_tests() {
        return {
            {"Single Element", {{5}}, 5, {5}},
            {"Two Rows", {{1}, {2, 3}}, 3, {1, 2}},
            {"All Same Values", {{1}, {1, 1}, {1, 1, 1}}, 3, {1, 1, 1}},
            {"Negative Values", {{-1}, {-2, -3}, {-4, -5, -6}}, -10, {-1, -3, -6}}
        };
    }
    
    static std::vector<TestCase> get_large_tests() {
        return {
            {"5x5 Triangle", {{1}, {2, 3}, {4, 5, 6}, {7, 8, 9, 10}, {11, 12, 13, 14, 15}}, 
             1 + 2 + 4 + 7 + 11, {1, 2, 4, 7, 11}}
        };
    }
};

struct TestResult {
    std::string name;
    bool passed;
    int actual_sum;
    std::vector<int> actual_path;
};

TestResult run_test(const TestCase& test_case, int test_number, Logger& logger) {
    logger.info(LogMessages::TEST_START + std::to_string(test_number));
    
    auto [actual_sum, actual_path] = minimum_total(test_case.triangle, logger);
    
    bool sum_correct = (actual_sum == test_case.expected_sum);
    bool path_correct = (actual_path == test_case.expected_path);
    bool passed = sum_correct && path_correct;
    
    if (passed) {
        logger.info(LogMessages::TEST_PASSED);
    } else {
        logger.info(LogMessages::TEST_FAILED);
    }
    
    std::string result_msg = LogMessages::TEST_RESULT + 
        std::to_string(test_number) + ": Expected sum = " + std::to_string(test_case.expected_sum) +
        ", Got = " + std::to_string(actual_sum) + " | Expected path = " + 
        pathToString(test_case.expected_path) + ", Got = " + pathToString(actual_path);
    logger.info(result_msg);
    
    return {test_case.name, passed, actual_sum, actual_path};
}

std::vector<TestResult> run_test_suite(Logger& logger) {
    logger.info("RUNNING COMPREHENSIVE TEST SUITE");
    
    std::vector<TestResult> results;
    std::vector<TestCase> all_tests;
    
    auto basic_tests = TriangleTests::get_basic_tests();
    all_tests.insert(all_tests.end(), basic_tests.begin(), basic_tests.end());
    
    auto edge_tests = TriangleTests::get_edge_tests();
    all_tests.insert(all_tests.end(), edge_tests.begin(), edge_tests.end());
    
    auto large_tests = TriangleTests::get_large_tests();
    all_tests.insert(all_tests.end(), large_tests.begin(), large_tests.end());
    
    for (size_t i = 0; i < all_tests.size(); ++i) {
        results.push_back(run_test(all_tests[i], i + 1, logger));
    }
    
    TriangleGenerator generator;
    for (int i = 0; i < 3; ++i) {
        int rows = 3 + (i * 2); // 3, 5, 7 rows
        auto triangle = generator.generate_random_triangle(rows);
        
        auto [expected_sum, expected_path] = minimum_total(triangle, logger);
        
        TestCase random_test = {
            "Random Test " + std::to_string(i + 1),
            triangle,
            expected_sum,
            expected_path
        };
        
        results.push_back(run_test(random_test, all_tests.size() + i + 1, logger));
    }
    
    return results;
}

void print_test_summary(const std::vector<TestResult>& results, Logger& logger) {
    logger.info("TEST SUMMARY");
    
    int passed_count = 0;
    for (const auto& result : results) {
        if (result.passed) passed_count++;
    }
    
    int total_count = results.size();
    
    logger.info("Total Tests: " + std::to_string(total_count));
    logger.info("Passed: " + std::to_string(passed_count));
    logger.info("Failed: " + std::to_string(total_count - passed_count));
    logger.info("Success Rate: " + std::to_string(static_cast<double>(passed_count) / total_count * 100) + "%");
    
    logger.info("Detailed Results:");
    for (const auto& result : results) {
        std::string status = result.passed ? "PASS" : "FAIL";
        logger.info(result.name + " " + status + " Sum: " + std::to_string(result.actual_sum) + 
                   " Path: " + pathToString(result.actual_path));
    }
}

void benchmark_algorithm(Logger& logger) {
    logger.info("BENCHMARK WITH LARGE TRIANGLES");
    
    TriangleGenerator generator;
    std::vector<int> sizes = {10, 20, 50, 100};
    
    for (int size : sizes) {
        auto triangle = generator.generate_random_triangle(size, -100, 100);
        
        auto start = std::chrono::high_resolution_clock::now();
        auto [min_sum, path] = minimum_total(triangle, logger);
        auto end = std::chrono::high_resolution_clock::now();
        
        auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
        double execution_time = duration.count() / 1000000.0;
        
        logger.info("Size: " + std::to_string(size) + ", Time: " + 
                   std::to_string(execution_time) + "s, Min Sum: " + std::to_string(min_sum));
    }
}

int main() {
    try {
        Logger logger;
        
        auto test_results = run_test_suite(logger);
        
        print_test_summary(test_results, logger);
        
        logger.info("BENCHMARK ... ");
        benchmark_algorithm(logger);
        
        auto original_tests = TriangleTests::get_basic_tests();
        for (const auto& test : original_tests) {
            auto [min_sum, path] = minimum_total(test.triangle, logger);
        }
        
    } catch (const std::exception& error) {
        std::cerr << "Error in main execution: " << error.what() << std::endl;
        return 1;
    }
    
    return 0;
}
